packages/graph/hash_graph/lib/graph/src/store/postgres/mod.rs

```diff
@@ -653,80 +685,219 @@ impl PostgresStore<Transaction<'_>> {
+            .into_iter()
+            .map(|row| row.get::<_, i64>(0))
+            .collect();
+
+        self.client
+            .simple_query("DROP TABLE entity_editions_temp;")
+            .await
+            .into_report()
+            .change_context(InsertionError)?;
+
+        Ok(entity_edition_ids)
+    }
+
+    #[doc(hidden)]
+    #[cfg(feature = "__internal_bench")]
+    async fn insert_entity_versions(
+        &self,
+        entities: impl IntoIterator<Item = (EntityId, i64, Option<DecisionTimestamp>), IntoIter: Send>
+        + Send,
+    ) -> Result<Vec<EntityVersion>, InsertionError> {
+        self.client
+            .simple_query(
+                "CREATE TEMPORARY TABLE entity_versions_temp (
+                    owned_by_id UUID NOT NULL,
+                    entity_uuid UUID NOT NULL,
+                    entity_edition_id BIGINT NOT NULL,
+                    decision_time TIMESTAMP WITH TIME ZONE
+                );",
+            )
+            .await
+            .into_report()
+            .change_context(InsertionError)?;
+
+        let sink = self
+            .client
+            .copy_in(
+                "COPY entity_versions_temp (
+                    owned_by_id,
+                    entity_uuid,
+                    entity_edition_id,
+                    decision_time
+                ) FROM STDIN BINARY",
+            )
+            .await
+            .into_report()
+            .change_context(InsertionError)?;
+        let writer = BinaryCopyInWriter::new(sink, &[
+            Type::UUID,
+            Type::UUID,
+            Type::INT8,
+            Type::TIMESTAMPTZ,
+        ]);
+        futures::pin_mut!(writer);
+        for (entity_id, entity_edition_id, decision_time) in entities {
+            writer
+                .as_mut()
+                .write(&[
+                    &entity_id.owned_by_id(),
+                    &entity_id.entity_uuid(),
+                    &entity_edition_id,
+                    &decision_time,
+                ])
+                .await
+                .into_report()
+                .change_context(InsertionError)?;
+        }
+
+        writer
+            .finish()
+            .await
+            .into_report()
+            .change_context(InsertionError)?;
+
+        let entity_versions = self
+            .client
+            .query(
+                "INSERT INTO entity_versions (
+                    owned_by_id,
+                    entity_uuid,
+                    entity_edition_id,
+                    decision_time,
+                    transaction_time
+                ) SELECT
+                    owned_by_id,
+                    entity_uuid,
+                    entity_edition_id,
+                    tstzrange(
+                        CASE WHEN decision_time IS NULL THEN now() ELSE decision_time END,
+                        'infinity'
+                    ),
+                    tstzrange(now(), 'infinity')
```

Comment:
nit: We can add the explicit bounds here
```suggestion
                        'infinity',
                        '[)'
                    ),
                    tstzrange(now(), 'infinity', '[)')
```